File: backend/healthcheck.py
#!/usr/bin/env python3
"""
Health check script for Docker container.

This script checks the health of the Django backend by
making a request to the dedicated health endpoint.
"""

import http.client
import json
import sys
import time
import os

def perform_health_check():
    """Perform health check and exit with appropriate status."""
    
    # Configuration
    hostname = "localhost"
    port = int(os.environ.get("PORT", 8000))
    path = "/health/"
    timeout = 3  # seconds
    
    try:
        # Connect with timeout
        start_time = time.time()
        conn = http.client.HTTPConnection(hostname, port, timeout=timeout)
        conn.request("GET", path, headers={"Accept": "application/json"})
        
        # Get response
        response = conn.getresponse()
        response_time = time.time() - start_time
        
        # Check status code
        if response.status == 200:
            # Parse response
            data = json.loads(response.read().decode("utf-8"))
            
            # Check if database is healthy
            if data.get("status") == "healthy" and data.get("database", {}).get("status") == "healthy":
                print(f"✅ Health check passed in {response_time:.3f}s - Database connected")
                sys.exit(0)
            else:
                print(f"❌ Health check failed - Status: {data.get('status')}, DB: {data.get('database', {}).get('status')}")
                sys.exit(1)
        else:
            print(f"❌ Health check failed with status code: {response.status}")
            sys.exit(1)
            
    except Exception as e:
        print(f"❌ Health check failed: {str(e)}")
        sys.exit(1)
    finally:
        try:
            conn.close()
        except:
            pass

if __name__ == "__main__":
    perform_health_check()

File: ssr/healthcheck.js
/**
 * Health check script for Docker container
 *
 * This script checks the health of the Next.js SSR service by
 * making a request to the dedicated health endpoint.
 */

const http = require("http");
const os = require("os");

// Configuration
const options = {
  hostname: os.hostname() || "localhost",
  port: parseInt(process.env.PORT) || 3000,
  path: "/health",
  method: "GET",
  timeout: 3000,
  headers: {
    Accept: "application/json",
  },
};

// Execute health check
const req = http.request(options, (res) => {
  let data = "";

  // Collect response data
  res.on("data", (chunk) => {
    data += chunk;
  });

  // Process the complete response
  res.on("end", () => {
    if (res.statusCode === 200) {
      try {
        // Parse the JSON response
        const healthData = JSON.parse(data);

        // Check if the status is healthy and API is connected
        if (
          healthData.status === "healthy" &&
          healthData.api?.status === "connected"
        ) {
          console.log("✅ Health check passed - API connection verified");
          process.exit(0);
        } else {
          console.error(
            `❌ Health check failed - Status: ${healthData.status}, API: ${healthData.api?.status}`,
          );
          process.exit(1);
        }
      } catch (e) {
        console.error(
          `❌ Health check failed - Invalid JSON response: ${e.message}`,
        );
        process.exit(1);
      }
    } else {
      console.error(`❌ Health check failed with status: ${res.statusCode}`);
      process.exit(1);
    }
  });
});

// Handle connection errors
req.on("error", (err) => {
  console.error(`❌ Health check failed: ${err.message}`);
  process.exit(1);
});

// Handle timeouts
req.on("timeout", () => {
  console.error("❌ Health check timed out");
  req.destroy();
  process.exit(1);
});

// Send the request
req.end();
